optimade/adapters/structures/cif.py

```diff
@@ -0,0 +1,129 @@
+import re
+from typing import Dict
+
+from optimade.models import Species as OptimadeStructureSpecies
+from optimade.models import StructureResource as OptimadeStructure
+
+from optimade.adapters.structures.utils import cell_to_cellpar, pad_positions
+
+try:
+    import numpy as np
+except ImportError:
+    from warnings import warn
+
+    np = None
+    NUMPY_NOT_FOUND = "NumPy not found, cannot convert structure to CIF"
+
+
+__all__ = ("get_cif",)
+
+
+def get_cif(  # pylint: disable=too-many-locals,too-many-branches
+    optimade_structure: OptimadeStructure,
+) -> str:
+    """ Get CIF file as string from OPTIMADE structure
+
+    Based on `ase.io.cif:write_cif()`.
+
+    :param optimade_structure: OPTIMADE structure
+    :param formatting: What formatting to use for the CIF file data keys.
+        Can be either "mp" or "default".
+    :param encoding: Encoding used for the string. CIF files use "latin-1" as standard.
+        If encoding is "str", a Python str object will be returned.
+    :return: str
+    """
+    # NumPy is needed for calculations
+    if globals().get("np", None) is None:
+        warn(NUMPY_NOT_FOUND)
+        return None
+
+    cif = """#
+# Created from an OPTIMADE structure.
+#
+# See https://www.optimade.org and/or
+# https://github.com/Materials-Consortia/OPTIMADE for more information.
+#
+"""
+
+    cif += f"data_{optimade_structure.id}\n\n"
+
+    attributes = optimade_structure.attributes
+
+    # Do this only if there's three non-zero lattice vectors
+    # NOTE: This also negates handling of lattice_vectors with null/None values
+    if sum(attributes.dimension_types) == 3:
+        a_vector, b_vector, c_vector, alpha, beta, gamma = cell_to_cellpar(
+            attributes.lattice_vectors
+        )
+
+        cif += (
+            f"_cell_length_a                    {a_vector:g}\n"
+            f"_cell_length_b                    {b_vector:g}\n"
+            f"_cell_length_c                    {c_vector:g}\n"
+            f"_cell_angle_alpha                 {alpha:g}\n"
+            f"_cell_angle_beta                  {beta:g}\n"
+            f"_cell_angle_gamma                 {gamma:g}\n\n"
+        )
+        cif += (
+            "_symmetry_space_group_name_H-M    'P 1'\n"
+            "_symmetry_int_tables_number       1\n\n"
+            "loop_\n"
+            "  _symmetry_equiv_pos_as_xyz\n"
+            "  'x, y, z'\n\n"
+        )
+
+    # Is it a periodic system?
+    # NOTE: This is a bit ahead of its time, since this OPTIMADE property is part of an open PR.
+    # See https://github.com/Materials-Consortia/OPTIMADE/pull/206
+    coord_type = (
+        "fract" if hasattr(attributes, "fractional_site_positions") else "Cartn"
+    )
```

Comment:
To summarise our slack discussion, I think the best thing to do is convert to fractionals here if a cell is provided, and if not just output the cartesian positions as you would for a molecule

Reply:
Actually not even sure we can write a cif file without a cell... maybe we just raise an error?

Reply:
> Actually not even sure we can write a cif file without a cell... maybe we just raise an error?

I can't see why we shouldn't be able to do this. It will instead be a problem for the viewer in question whether they can handle cif files without cell information or not.

However, for the other issue, I agree (as well from the slack discussion) that we should convert to fractional coordinates if we have sufficient information to do so and otherwise convert to a "molecule-style" output.

Reply:
Actually, I am already _not_ including cell information in the cif file if we are not dealing with a 3D structure.